### ISSUE TITLE
CI and scripts improvements

### ISSUE THREAD

**Issue by darosh** [APIs-guru/openapi-directory#334]
I noticed the space for improvements in the scripts and I have some good news: rebuild takes [**2+ minutes**](https://travis-ci.org/darosh/openapi-directory/builds/306878936) instead of [**10+ minutes**](https://travis-ci.org/APIs-guru/openapi-directory/builds/305928798). It took a lot of ☕️s and I am excited it works!

There is more concurrency (for git logs, logo downloads), caching (logos, hashed validation result, S3 status) leveraging Travis cache.

Work in progress here: https://github.com/darosh/openapi-directory/tree/scripts/gulp and here: https://github.com/darosh/openapi-directory/blob/scripts/gulpfile.babel.js

Update `collection.js update` (still WIP) is about 5 minutes, turned into gulp pipeline:

![image](https://user-images.githubusercontent.com/297320/33221820-e3a29e96-d152-11e7-871e-b5733ac36db7.png)

Notice: max concurrency numbers for async tasks in picture above. Most of code is unchanged, Gulp's vinil file system is used as "context" (`ctx` , sometimes deconstructed in most functions in https://github.com/darosh/openapi-directory/tree/scripts/gulp/lib/spec ... it is sort of refactoring from imperative to functional, putting the 'imperative' into the pipeline) The idea of individual specs streams is similar to other linting tools: don't halt/assert on first error, collect errors for each, stop fatal stream processing, but do not break whole, report them all...

easy to pipe any intermediate state into file-based log (notice the folder structure in picture bellow):

![image](https://user-images.githubusercontent.com/297320/33221879-509c7012-d153-11e7-81b9-49778abeab18.png)

Currently (~80%) implemented Gulp interface:

![image](https://user-images.githubusercontent.com/297320/33221962-e3972bbe-d153-11e7-94d1-25c33a8f0201.png)

There is still a lot of work to be done. Looking fwd for your feedback. More to come. Many questions as well. 😸

**Comment by MikeRalphson** [comment]
Wow! That looks like a lot of work. I must confess I'm completely unfamiliar with gulp so it will take a lot of explaining (and maybe convincing)... The time for CI to run really isn't a major issue, but if the update process could be made more smooth, performant and robust, they would be worthwhile steps forward.

Pinging @IvanGoncharov

**Comment by darosh** [comment]
@MikeRalphson

> _unfamiliar with gulp so it will take a lot of explaining_

[Gulp.js](https://github.com/gulpjs/gulp) is task runner build around node streams encapsulated in [virtual file format](https://github.com/gulpjs/vinyl) (aka vinyl) objects made for concurrency with simple API.

`src`, `pipe`, `dest`, `series`, `parallel,` `log`, `File`, `PluginError` thats pretty all. 

It feels like right tool for such job a the logging of each step in the update pipeline helped me a lot.

> _time for CI to run really isn't a major issue_ 

I started with the simple tasks to come with something functional first. 

It will be easier to extend build with some additional features, for example zipped archive, build timestamp, smaller index (like the [one](https://darosh.github.io/openapi-directory-lite/index.json) I made), ...

> _more smooth, performant and robust_ 

Yes. And more testable, configurable, with better processing logs and reports, and file based logs.

> _Ditching Bluebird..._ 

👍 I replaced Bluebird specific `each`, `done`, `spread` where I spotted it automatically 😀  ... as far as I know it is required only in in the [converter http request](https://github.com/darosh/openapi-directory/blob/scripts/gulp/lib/converter.js#L4) because they use `spread` 😞. 

Side notes to converter, they could remove:
- `spread` allowing native Promises
- `CompositeError` (throws another exception on exception serialization attempt)
- every spec processing starts with `existsSync`, even for URLs

Additionally it looks like a lot of `then().then().then().then()..`. chains can be replaced with simple calls of extracted functions. (I guess engine adds new microtask to queue for each `then` ...which feels so 🤔 _...nondeterministic_) 

I want to mark [spec exports](https://github.com/darosh/openapi-directory/blob/scripts/gulp/lib/spec/index.js) with `Async` suffix first and use promises only where they are really needed.

Do you use all tasks in `collection.js`?

```
  Usage: collection [options] [command]


  Options:

    -0  always return 0 as exit code
    -h, --help  output usage information


  Commands:

    urls                          show source url for definitions
    refresh [DIR]                 run refresh - read and write back fixup.yaml files
    fixup <Swagger>               update "fixup.yaml" for specified Swagger
    update [options] [DIR]        run update
    validate [options] [DIR]      validate collection
    check [DIR]                   check status of x-preferred flags only
    leads                         add/remove definitions from 3rd-party catalogs
    add [options] <FORMAT> <URL>  add new definition
```

**Comment by MikeRalphson** [comment]
I don't think I've ever used `urls` but I've used every one of the other `collection.js` commands **today**.

**Comment by MikeRalphson** [comment]
When I said unfamiliar with Gulp I just meant I was unfamiliar with *using it* (I know roughly what it does) hence the need to explain the details of the changes so I can make an informed decision about migrating to Gulp tasks for everything.

If you're considering a PR, I'd prefer a starting point on a new branch which **does not** add any additional functionality, just improves either the CI task or (say) the `update` task. That way I can test it in parallel for some time before making a change.

**Comment by darosh** [comment]
I see, _explain the details of the changes_ makes perfect sense, hoping such [refactoring notes](https://github.com/darosh/openapi-directory/blob/scripts/gulp/README.md) (in progress) will help.

Regarding PR: understood, working in `scripts` branch of fork already, no changed functionality is exactly what refactoring should achieve, original scripts and gulp should produce same output for easy side by side deep folder compare.

**Comment by IvanGoncharov** [comment]
@darosh Very cool 👍 
My scripts are the big pile of mess in their current state so it's good if some refactoring is done.
I don't know much about Gulp but concepts of streams and Vinyl sims very appropriate.
I especially like the idea of partial failing it can enable some exciting ideas around CRON jobs and different bots.
For example, using Travis cron feature to run an update and submit PR with updated files and error messages (if any) inside PR text.

As for bluebird I used it because of `BLUEBIRD_LONG_STACK_TRACES=1 BLUEBIRD_WARNINGS=1`
http://bluebirdjs.com/docs/api/promise.config.html
It helped a lot with debugging at that time, I would suggest async/await as a natural step forward.

But most importantly refactoring and splitting everything into small chanks should help with external contribution. And allow adding new features easily, for example, separate script to validate all HTTP/HTTPS urls inside Swagger file. Or deploying specs as NPM package which should be helpful for testing Swagger tools against entire collection.

As a next step forward, I would suggest adding testing script between building folder and deploying it to S3. Something similar to what we discussed here: https://github.com/APIs-guru/openapi-directory/commit/12060c3539b97142bd18051c9397e8f7c9de4366#commitcomment-25601401
And then new version could be merged into repo without risk of breaking API.

I think it's critical to merge the first version before going forward with additional refactoring or new features.

BTW. Since I rename repo to `openapi-directory` how about renaming CLI command to `directory`?
It will allow having two commands alongside until migration completes.

**Comment by darosh** [comment]
Hi guys, Thanks a lot for your comments! This is quite challenging (as you said _pile of mess_ 😄) 

Right now I am looking at update process and the `getSwaggerPath` (based on version, service, provider) and I have few slightly related questions:

1) azure collection is looking at `master` branch which says `This branch is 11 commits ahead, 461 commits behind current.`, and they use `current` branch as default https://github.com/Azure/azure-rest-api-specs . Is `master` obsolete and should we use `current` instead?


2) This may be a minor issue or not issue at all because it is marked as `x-apisguru-direct`, I made a simple `path integrity` check like this:

```javascript
export function validateRelativePath (field) {
  return function (ctx) {
    const realPath = normalizePath(ctx.relative)
    const specPath = getSwaggerInfoPath(ctx[field].info)
    assert(realPath === specPath, `invalid spec path "${specPath}"`)
  }
}
```

gives me:

```
[18:50:00] [error] invalid spec path "amazonaws.com/autoscaling/2016-02-06/swagger.yaml" amazonaws.com\application-autoscaling\2016-02-06\swagger.yaml
[18:50:00] [error] invalid spec path "amazonaws.com/appstream2/2016-12-01/swagger.yaml" amazonaws.com\appstream\2016-12-01\swagger.yaml
[18:50:00] [error] invalid spec path "amazonaws.com/mgh/2017-05-31/swagger.yaml" amazonaws.com\AWSMigrationHub\2017-05-31\swagger.yaml
[18:50:00] [error] invalid spec path "amazonaws.com/elasticloadbalancing/2015-12-01/swagger.yaml" amazonaws.com\elasticloadbalancingv2\2015-12-01\swagger.yaml
[18:50:01] [error] invalid spec path "amazonaws.com/data.iot/2015-05-28/swagger.yaml" amazonaws.com\iot-data\2015-05-28\swagger.yaml
[18:50:01] [error] invalid spec path "amazonaws.com/models.lex/2017-04-19/swagger.yaml" amazonaws.com\lex-models\2017-04-19\swagger.yaml
[18:50:01] [error] invalid spec path "amazonaws.com/metering.marketplace/2016-01-14/swagger.yaml" amazonaws.com\meteringmarketplace\2016-01-14\swagger.yaml
[18:50:01] [error] invalid spec path "amazonaws.com/opsworks-cm/2016-11-01/swagger.yaml" amazonaws.com\opsworkscm\2016-11-01\swagger.yaml
[18:50:01] [error] invalid spec path "amazonaws.com/api.pricing/2017-10-15/swagger.yaml" amazonaws.com\pricing\2017-10-15\swagger.yaml
[18:50:01] [error] invalid spec path "amazonaws.com/tagging/2017-01-26/swagger.yaml" amazonaws.com\resourcegroupstaggingapi\2017-01-26\swagger.yaml
```

It means only reported files are in different folder than expected because of different `x-serviceName` field.

**Comment by MikeRalphson** [comment]
@darosh re: the second point, yes the AWS APIs are basically just copied into place as one (the `x-apisguru-direct` flag) so the converter can see multiple versions of each API at once. Amazon are a little inconsistent between the naming of their files and the data element we use as the `serviceName`. I can correct this quite easily by deciding on one or the other.

Once your work is integrated it will be a good test for me if I can come up with a Gulp task to pull in the AWS APIs, it will help me understand the process to work on that.

**Comment by MikeRalphson** [comment]
@darosh I am looking at switching the Azure branch to `current` now, but it is going to involve some extra handling of the filenames. Thanks for pointing this out!

**Comment by MikeRalphson** [comment]
@darosh thanks also for noting the Amazon naming inconsistencies. They should be resolved now.

**Comment by MikeRalphson** [comment]
@darosh Hi, do you have any update on this? Obviously as time goes on the original scripts are evolving and may now differ from what you started with.

**Comment by MikeRalphson** [comment]
@darosh are you still wanting to work on this, or should I close it?

**Comment by MikeRalphson** [comment]
Closing due to inactivity - but please feel free to reopen the issue if necessary.
